index.js

```diff
@@ -0,0 +1,254 @@
+ * dependencies between chunks (via the 'parents' property on chunks) based on
+ * whether a module already exists in shared chunk or not.
+ *
+ * Shared chunks are created in the order they are specified.
+ *
+ * A set of shared chunks can be specified in a webpack.config.js like so:
+ *
+ *  plugins: [
+ *      new SharedChunkPlugin({
+ *          specs: [
+ *              {
+ *                  name: "vendor",
+ *                  selectedChunks: ["foo", "bar", "baz", "qux"],
+ *                  moduleFilter: (module) =>
+ *                      /vendor/.test(module.resource),
+ *              },
+ *              {
+ *                  name: "components",
+ *                  selectedChunks: ["foo", "bar", "baz", "qux"],
+ *                  moduleFilter: (module) =>
+ *                      /components/.test(module.resource),
+ *              },
+ *              {
+ *                  name: "feature-shared",
+ *                  selectedChunks: ["foo", "bar", "baz", "qux"],
+ *                  moduleFilter: (module) =>
+ *                      /features/.test(module.resource),
+ *                  minChunks: 2,
+ *              },
+ *          ]
+ *      })
+ *  ],
+ *
+ * Each of the 'specs' can contain the following options:
+ * - name: the name of the chunk being generated
+ * - selectedChunks: an array of entry chunks to consider when looking for
+ *                   common modules
+ * - moduleFilter: a predicate which can be used to determine whether a module
+ *                 should be included in the shared chunk that is currently
+ *                 being created
+ * - minChunks (optional): a number between 1 and Infinity that specifies how
+ *                         many entry chunks it must appear in before it is
+ *                         extracted to the shared chunk.
+ */
+class SharedChunkPlugin {
+    constructor(options) {
+        this.specs = options.specs;
+    }
+
+    apply(compiler) {
+        compiler.plugin("compilation", (compilation) => {
+            compilation.plugin(["optimize-chunks"], (allChunks) => {
+                const globalModulesSet = new Set();
+                for (const spec of this.specs) {
+                    this.processSharedSpec(
+                        compilation, allChunks, globalModulesSet, spec);
+                }
+            });
+
+            // additional-chunk-assets is the only unconditional hook that gets
+            // passed all of the chunks that is called after sortItemsWithModuleIds
+            // and sortItemsWithChunkIds which are two internal webpack methods
+            // in Compilation.js that sort the order of the parents based on
+            // chunk id.
+            compilation.plugin(["additional-chunk-assets"], (allChunks) => {
+                for (const chunk of allChunks) {
+                    // Sort dependencies based on the order in which they should
+                    // be loaded.
+                    const sortFn = (a, b) => {
```

Comment:
Defining this inside the for loop doesn't feel right to me, should be a static function.